geo/src/algorithm/concave_hull.rs

```diff
@@ -87,363 +130,536 @@ where
     T: GeoFloat + RTreeNum,
 {
     type Scalar = T;
-    fn concave_hull(&self, concavity: T) -> Polygon<T> {
-        let mut aggregated: Vec<Coord<T>> = self.iter().flat_map(|elem| elem.0.clone()).collect();
-        Polygon::new(concave_hull(&mut aggregated, concavity), vec![])
+    /// Note that the concave hull may intersect with the original geometry boundaries.
+    fn concave_hull(&self, concavity: Self::Scalar, length_threshold: Self::Scalar) -> Polygon<T> {
+        let mut coords: Vec<Coord<T>> = self.iter().flat_map(|elem| elem.0.clone()).collect();
+        concave_hull(&mut coords, concavity, length_threshold)
     }
 }
 
-impl<T> ConcaveHull for MultiPoint<T>
+impl<T> ConcaveHull for Vec<Coord<T>>
 where
     T: GeoFloat + RTreeNum,
 {
     type Scalar = T;
-    fn concave_hull(&self, concavity: T) -> Polygon<T> {
-        let mut coordinates: Vec<Coord<T>> = self.iter().map(|point| point.0).collect();
-        Polygon::new(concave_hull(&mut coordinates, concavity), vec![])
+    fn concave_hull(
+        &self,
+        concavity: Self::Scalar,
+        length_threshold: Self::Scalar,
+    ) -> Polygon<Self::Scalar> {
+        let mut coords: Vec<Coord<T>> = self.clone();
+        concave_hull(&mut coords, concavity, length_threshold)
     }
 }
 
-fn find_point_closest_to_line<T>(
-    interior_coords_tree: &RTree<Coord<T>>,
+impl<T> ConcaveHull for [Coord<T>]
+where
+    T: GeoFloat + RTreeNum,
+{
+    type Scalar = T;
+    fn concave_hull(
+        &self,
+        concavity: Self::Scalar,
+        length_threshold: Self::Scalar,
+    ) -> Polygon<Self::Scalar> {
+        let mut coords: Vec<Coord<T>> = self.to_vec();
+        concave_hull(&mut coords, concavity, length_threshold)
+    }
+}
+
+enum RTreeNodeRef<'a, T>
+where
+    T: GeoFloat + RTreeNum,
+{
+    Parent(&'a ParentNode<Coord<T>>),
+    Leaf(&'a Coord<T>),
+}
+
+struct NodeQueueItem<'a, T>
+where
+    T: GeoFloat + RTreeNum,
+{
+    tree_node: RTreeNodeRef<'a, T>,
+    distance: T,
+}
+
+impl<'a, T> Ord for NodeQueueItem<'a, T>
+where
+    T: GeoFloat + RTreeNum,
+{
+    fn cmp(&self, other: &Self) -> Ordering {
+        other.distance.partial_cmp(&self.distance).unwrap()
+    }
+}
+impl<'a, T> PartialOrd for NodeQueueItem<'a, T>
+where
+    T: GeoFloat + RTreeNum,
+{
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+impl<'a, T> PartialEq for NodeQueueItem<'a, T>
+where
+    T: GeoFloat + RTreeNum,
+{
+    fn eq(&self, other: &Self) -> bool {
+        self.distance == other.distance
+    }
+}
+impl<'a, T> Eq for NodeQueueItem<'a, T> where T: GeoFloat + RTreeNum {}
+
+struct LineQueueItem<T: GeoFloat + RTreeNum> {
     line: Line<T>,
-    max_dist: T,
-    edge_length: T,
-    concavity: T,
-    line_tree: &RTree<Line<T>>,
+    i: usize,
+    prev_i: usize,
+    next_i: usize,
+}
+
+fn line_to_bbox_distance<T>(line: &Line<T>, aabb: &AABB<Coord<T>>) -> T
```

Comment:
It's not immediately obvious what this function is for (I can take a pretty good guess, but I'd prefer not to). Could you document its purpose?